content/blogposts/2024/anatomy-of-a-coding-assistant.md

```diff
@@ -0,0 +1,157 @@
+
+In this post, we're going to take a look at how context fetching works for each particular feature - chat, autocomplete, generating tests, and editing code.
+
+---
+
+## How context fetching works for chat
+
+![8_chat](https://storage.googleapis.com/sourcegraph-assets/blog/anatomy/8_chat.png)
+
+Chat allows you to ask questions about your codebase and generate code that's tailored to work within your codebase. For this, we want a few things:
+
+1. **Conversation history:** In a given chat session, we record previous messages as they may contain relevant context for the user's next request.
+2. **Code search:** We fetch the most relevant code snippets related to the user's query from the codebase using search, similar to how a human dev might search for these code snippets.
+3. **User control:** Users should have the ability to mention specific files and provide those directly to the model with an option to reference external sources like Slack threads or Notion documents to enrich the context further.
+
+![9_context](https://storage.googleapis.com/sourcegraph-assets/blog/anatomy/9_context.png)
+
+For code search, we use two techniques for surfacing relevant snippets:
+
+1. **AI-enhanced keyword search:** We ask an LLM a question like, "Given this user query, what keywords should be used to search for relevant parts of the codebase?" We use these keywords to query a keyword search index we've created over the codebase. This index returns a set of results, ordered by Sourcegraph's code-aware ranking mechanism, and we select the top results subject to the constraint of the LLM's context window.
+    
+![10_indexed](https://storage.googleapis.com/sourcegraph-assets/blog/anatomy/10_indexed.png)
+    
+2. **Embeddings:** Embeddings refer to the internal vector of numbers that LLMs use to represent text. When we set up embeddings, we first convert code snippets from various files into embeddings in an indexing step. At query time, we convert the user query to an embedding vector, too. Using these embeddings, we perform vector search to find the most relevant code snippets for the user query.
+    1. To convert code snippets into embeddings, we use either OpenAI’s text-embedding-ada-002 model or Sourcegraph's own model called st-multi-qa-mpnet-base-dot-v1. The choice between these models is [determined by a feature flag](https://github.com/sourcegraph/cody/blob/ee47055967bdb3e65c8694cab726dd47d50c29ba/vscode/src/local-context/local-embeddings.ts#L31). The embeddings are [stored locally on your machine](https://github.com/sourcegraph/cody/blob/main/vscode/src/local-context/local-embeddings.ts#L55).
+    
+![11_embeddings](https://storage.googleapis.com/sourcegraph-assets/blog/anatomy/11_embeddings.png)
+    
+
+Currently, [we default to a mix of both](https://github.com/sourcegraph/cody/blob/0238df0eda7e6e9be03139a56a89e235b76be448/vscode/package.json#L967) with an option to use one or the other.
+
+### Recent improvements to context fetching for chat
```

Comment:
What if we just removed this section, in favor of mentioning these things under the "Keyword Search" section, as my comment above suggests? Maybe it's just me, but these improvements don't sound that technically impressive on their own :)